modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java

```diff
@@ -717,4 +724,59 @@ private CompletableFuture<Void> closeExecNode(boolean cancel) {
     interface ImplementorFactory<RowT> {
         LogicalRelImplementor<RowT> create(ExecutionContext<RowT> ctx);
     }
+
+    private static class ColocationServiceImpl<T> implements ColocationHashFunctionFactory<T> {
+        private final SqlSchemaManager sqlSchemaManager;
+        private final RowHandler<T> rowHandler;
+
+        private ColocationServiceImpl(SqlSchemaManager sqlSchemaManager, RowHandler<T> rowHandler) {
+            this.sqlSchemaManager = sqlSchemaManager;
+            this.rowHandler = rowHandler;
+        }
+
+        @Override
+        public ColocationHashFunction<T> create(UUID tableId, int[] colocationFields) {
+            return new ColocationHashFunctionImpl(tableId, colocationFields);
+        }
+
+        private class ColocationHashFunctionImpl implements ColocationHashFunction<T> {
+            private final int[] colocationFields;
+            private final NativeType[] colTypes;
+
+            private ColocationHashFunctionImpl(UUID tableId, int[] colocationFields) {
+                this.colTypes = colocationColumnTypes(tableId, colocationFields.length);
+                this.colocationFields = colocationFields;
+            }
+
+            private NativeType[] colocationColumnTypes(UUID tableId, int fieldCnt) {
+                NativeType[] colTypes = new NativeType[fieldCnt];
+                TableDescriptor tblDesc = sqlSchemaManager.tableById(tableId, -1).descriptor();
+
+                List<Integer> colocationColumns = tblDesc.colocationColumns();
+
+                assert colocationColumns.size() >= fieldCnt : "fieldsCount=" + fieldCnt + ", colocationColumns=" + colocationColumns;
```

Comment:
How it's going to work if table colocated by more column than we expected?

Reply:
Fixed, thanks